client/network/src/protocol/generic_proto/behaviour.rs

```diff
@@ -1123,25 +1123,25 @@ impl NetworkBehaviour for GenericProto {
 							);
 						}
 
+						// TODO: We switch the entire peer state to "disabled" because of possible
+						// race conditions involving the legacy substream.
+						// Once https://github.com/paritytech/substrate/issues/5670 is done, this
+						// should be changed to stay in the `Enabled` state.
 						debug!(target: "sub-libp2p", "Handler({:?}) <= Disable", source);
+						debug!(target: "sub-libp2p", "PSM <= Dropped({:?})", source);
+						self.peerset.dropped(source.clone());
 						self.events.push(NetworkBehaviourAction::NotifyHandler {
 							peer_id: source.clone(),
-							handler: NotifyHandler::One(connection),
+							handler: NotifyHandler::All,
```

Comment:
You now disable all handlers, but note that `last` is still only true if the last handler reports `Closed`, and only if `last` is true will the code further below emit `CustomProtocolClosed`. Meaning you now always disable the peer but you do not always report `CustomProtocolClosed`.

Reply:
That is intended. Whatever the situation, we report `CustomProtocolClosed` when and only when `open` becomes empty (and similarly we report `CustomProtocolOpen` when `open` becomes non-empty).

Reply:
But disabling all handlers effectively means closing all connections (eventually), so `CustomProtocolClosed` must follow inevitably. You're saying you just want to wait for that to happen via `inject_disconnected`?

Reply:
Ok, I think I see what you mean.